HIP/hip-904-frictionless-airdrop.md

```diff
@@ -0,0 +1,711 @@
+    * then the payer will pay for the transfer, association, one renewal credit,
+    * and airdrop fee. If the receiver has not associated with the token and has
+    * no open association slot, then the transfer will not deliver the token,
+    * but will create a `pending airdrop` instead, with the receiver account
+    * listed on the `pending airdrop`. The payer will pay for the transfer,
+    * association, one renewal credit, and airdrop fee.
+    * All tokenTransferList items must succeed (transfer of pedning state) for the transaction to succeed.
+    */
+    repeated TokenTransferList tokenTransfers = 1;
+}
+```
+
+TokenAirdrop transaction will support the transfers of value in a similar manner to CryptoTransfer with the difference of only supporting tokens (no HBAR transfer support) and supporting the resulting pending transfer state where a CryptoTransfer would have failed due to the lack of association options.
+
+However, it is important to have a separate HAPI transaction type for a few reasons
+
+- the ability to call out the potential separate behaviours based on maxAutomaticAssociation, whereby transfer can end up in a pending transfer state or a consumption of an association slot.
+- cleaner conceptual understanding
+- naming allows the grouping of functionality with Cancel and Claim which complete the pending state a TokenAirdrop may creates
+- explicit specification in record stream.
+- improved developer experience
+
+Similar restrictions around transfer list and aggregation across sender and receiver from CryptoTransfer will be applied to TokenAirdop e.g. max 10 transfers per transaction
+
+At the time of this HIP `TokenAirdropTransaction` will not be supported by the Scheduled Transaction Service. A future HIP may explore this support.
+
+#### Cancel
+
+```protobuf
+message TokenPendingAirdrop {
+    /**
+    * Account that issue the airdrop trasnaction
+    */
+	AccountID sender_id = 1;
+
+	/**
+    * Account that was the intended reciever of the airdrop
+    */
+    AccountID receiver_id = 2;
+
+	/**
+    * The unique token identifier. If it's a FT then it's just the TokenID.
+	* If it's an NFT then NftID which captures both the TokenID and serial id
+    */
+	oneof token_id {
+		NftID non_fungible_token = 3;
+		TokenID fungible_token = 4;
+	}	
+}
+
+```
+
+```protobuf
+/**
+ * A token cancel will allow an airdrop sender to revert clear a pending transfer from state.
+ * Afer this no pending transfer will exist for a potential receiver to claim.  
+ * This transfer will be represented in the transaction body and will not need restating in the record file.
+ * All cancellations must succeed for the transaction to succeed.
+ */
+message TokenCancelAirdropTransactionBody {
+    /**
+    * A list of up to 10 pending airdrops that may be cancelled. If there are
+    * duplicate pending_airdrops, then the transaction will fail.
+    */
+    repeated TokenPendingAirdrop pending_airdrops = 1;
```

Comment:
Would it be more efficient to enable a list of receiver account ids in the TokenPendingAirdrop message, rather than repeat that message here ?
I'm thinking my proposal would result in smaller transaction sizes for the same outcome.

Reply:
For NFT you could just have a list of serial numbers (maximum length 10).  Each is unique, so the serial number is sufficient.
Using a OneOf here would make that fairly efficient
```
    oneof items_to_cancel {
        /**
        * A list of accounts with pending airdrops for a fungible/common token
        * that SHOULD be cancelled.<br/>
        * This list MUST NOT have duplicate entries.<br/>
        * This list MUST be 10 items or less.
        */
        repeated AccountID account_pending_airdrops = 1;

        /**
        * A list of serial numbers with pending airdrops for a non-fungible/unique
        * token that SHOULD be cancelled.<br/>
        * This list MUST NOT have duplicate entries.<br/>
        * This list MUST be 10 items or less.
        */
        repeated int64 serial_number_pending_airdrops = 2;
    }
```
This would change the record stream (we'd have to include the cancelled pending transfers there), but I suspect it would be clearer that way, and perhaps easier to process.

Reply:
Protobuf suggestions was re-reviewed with considerations for transaction I/O and state management and a balance was struck